### ISSUE TITLE
Improve opcode dispatch further by using a local curr_pc in executor

### ISSUE THREAD

**Issue by svaarala** [svaarala/duktape#294]
With #265 bytecode dispatch uses a cached `thr->curr_pc` pointer which is copied back into the topmost activation `act->curr_pc` whenever activation states may change or an up-to-date activation state is needed (e.g. for line numbers in the debugger).

This can probably be improved a bit further:

- Use a local variable in the bytecode executor for `curr_pc`. Make that variable volatile so that it is always stored back into the stack frame after a dispatch.
- Store a pointer to the stack frame location which holds `curr_pc` (i.e. a `duk_instr_t **` with `&curr_pc` as its value) in `thr->ptr_curr_pc`.
- Wherever `thr->curr_pc` was copied back into `act->curr_pc` after #265, copy the value from `*thr->ptr_curr_pc` instead.

This is faster because `thr` is sometimes spilled but the stack frame pointer is not.

Preliminary testing suggests this is roughly 5% faster than #265 (to be verified with more careful testing).

Changes:

- [x] Main source changes to implement the change
- [x] Perhaps simplify management of `thr->ptr_curr_pc`: bytecode executor can establish it, and restore it before exiting (it always has a setjmp catchpoint so it can do this reliably)
- [x] Check volatile declarations
- [x] Verify performance difference
- [x] Update doc/execution.rst
- [x] Testcases (apitest, qecmatest, test262)
- [x] Test debugger manually

**Comment by svaarala** [comment]
After #265 there seems to be a spurious compiler warning:

```
duk_js_call.c: In function ‘duk_handle_call’:
duk_js_call.c:835:15: warning: variable ‘entry_thr_curr_pc’ might be clobbered by ‘longjmp’ or ‘vfork’ [-Wclobbered]
```

See if that can be fixed with this change too.

**Comment by svaarala** [comment]
Quick test using a dispatch heavy test (LDCONST micro benchmark):

```
# Local curr_pc
$ sudo nice -20 python util/time_multi.py --count 10 --sleep 10 --mode all --verbose ./duk.O2.localpc tests/perf/test-const-load.js 
Running: 5.430000 5.410000 5.780000 5.370000 5.480000 5.560000 5.450000 5.380000 5.390000 5.370000
min=5.37, max=5.78, avg=5.46, count=10: [5.43, 5.41, 5.78, 5.37, 5.48, 5.56, 5.45, 5.38, 5.39, 5.37]

# Master, including thr->curr_pc dispatch and local consts base pointer
$ sudo nice -20 python util/time_multi.py --count 10 --sleep 10 --mode all --verbose ./duk.O2.master tests/perf/test-const-load.js 
Running: 6.430000 6.450000 6.440000 6.440000 6.420000 6.450000 6.410000 6.420000 7.190000 7.220000
min=6.41, max=7.22, avg=6.59, count=10: [6.43, 6.45, 6.44, 6.44, 6.42, 6.45, 6.41, 6.42, 7.19, 7.22]
```

Practical difference will be less, depending on how dispatch heavy the executed bytecode is.

**Comment by fatcerberus** [comment]
By "dispatch heavy" I assume you mean, e.g. tight loops?

**Comment by svaarala** [comment]
Anything that does a lot of opcode dispatch compared to other execution - if a tight loop involves a regexp match for instance, the regexp performance will of course dominate.

**Comment by svaarala** [comment]
I'll polish and merge this branch tomorrow. Performance work for 1.3 will then be finished, yay :-)

**Comment by svaarala** [comment]
C declarations are weird:

```
cdecl> explain int * volatile * volatile curr_pc
declare curr_pc as volatile pointer to volatile pointer to int
```

**Comment by fatcerberus** [comment]
I guess that's what happens when you're about half a step away from assembly. ;-)

**Comment by svaarala** [comment]
Quick performance test between master (thr->curr_pc and local curr_pc):

```
test-array-read.js            : duk-O2  4.00 duk-O2-master  4.31
test-array-write.js           : duk-O2  4.39 duk-O2-master  4.75
test-bitwise-ops.js           : duk-O2  0.87 duk-O2-master  0.94
test-buffer-nodejs-read.js    : duk-O2  4.41 duk-O2-master  4.84
test-buffer-nodejs-write.js   : duk-O2  5.56 duk-O2-master  6.15
test-buffer-object-read.js    : duk-O2  4.62 duk-O2-master  4.75
test-buffer-object-write.js   : duk-O2  5.78 duk-O2-master  5.94
test-buffer-plain-read.js     : duk-O2  3.97 duk-O2-master  4.50
test-buffer-plain-write.js    : duk-O2  3.64 duk-O2-master  4.17
test-call-basic.js            : duk-O2 15.08 duk-O2-master 16.04
test-compile-mandel-nofrac.js : duk-O2 13.90 duk-O2-master 13.59
test-compile-mandel.js        : duk-O2 17.27 duk-O2-master 17.01
test-compile-short.js         : duk-O2 10.42 duk-O2-master 10.84
test-const-load.js            : duk-O2  6.06 duk-O2-master  7.40
test-empty-loop.js            : duk-O2  2.18 duk-O2-master  2.33
test-fib.js                   : duk-O2  8.52 duk-O2-master  9.70
test-global-lookup.js         : duk-O2 11.66 duk-O2-master 10.42
test-hello-world.js           : duk-O2  0.00 duk-O2-master  0.00
test-hex-decode.js            : duk-O2  5.32 duk-O2-master  5.33
test-hex-encode.js            : duk-O2 26.94 duk-O2-master 27.30
test-json-integer-parse.js    : duk-O2  0.40 duk-O2-master  0.40
test-json-number-parse.js     : duk-O2  5.07 duk-O2-master  5.15
test-json-serialize.js        : duk-O2  2.97 duk-O2-master  3.00
test-json-string-bench.js     : duk-O2  5.59 duk-O2-master  5.50
test-json-string-parse.js     : duk-O2  0.59 duk-O2-master  0.59
test-json-string-stringify.js : duk-O2  1.27 duk-O2-master  1.27
test-prop-read.js             : duk-O2  8.01 duk-O2-master  7.96
test-prop-write.js            : duk-O2  7.18 duk-O2-master  7.17
test-reg-load.js              : duk-O2  2.76 duk-O2-master  3.12
test-reg-readwrite-object.js  : duk-O2  3.73 duk-O2-master  3.82
test-reg-readwrite-plain.js   : duk-O2  2.51 duk-O2-master  2.84
test-regexp-string-parse.js   : duk-O2 10.55 duk-O2-master 10.99
test-string-array-concat.js   : duk-O2  6.41 duk-O2-master  7.12
test-string-compare.js        : duk-O2  4.16 duk-O2-master  4.12
test-string-decodeuri.js      : duk-O2  3.67 duk-O2-master  3.63
test-string-encodeuri.js      : duk-O2  4.30 duk-O2-master  4.53
test-string-intern-match.js   : duk-O2  2.35 duk-O2-master  2.37
test-string-intern-miss.js    : duk-O2  2.54 duk-O2-master  2.56
test-string-plain-concat.js   : duk-O2  4.11 duk-O2-master  4.20
test-string-uppercase.js      : duk-O2  2.60 duk-O2-master  2.73
```

**Comment by svaarala** [comment]
This is now in master and is the final performance change for 1.3. There's potential for very weird fallout if something doesn't work right, so keep your eyes peeled :-)

**Comment by fatcerberus** [comment]
I'll pull the latest master and try it out with minisphere.  If there are any issues I'll let you know!

**Comment by fatcerberus** [comment]
Just got around to looking at the recent changes.  I must say this was quite clever:

```c
duk_instr_t **ptr_curr_pc;
```

The `duk_hthread` holds a backpointer to the executor's local PC to avoid an extra level of syncing (which is what I thought this change was going to entail).

Only one thing worries me about this setup, and that's how it will scale to recursive executor calls (as would happen in the case of, e.g. JS -> C -> JS.

**Comment by fatcerberus** [comment]
Well, no weird issues showed up in my (admittedly not very thorough) testing.  So it looks like this bit of black magic worked. ;-)

**Comment by svaarala** [comment]
Each time the bytecode executor is called recursively the call handling will store and restore `ptr_curr_pc` so it should not be an issue. That's not to say I'm not a bit uncomfortable with "syncing back", but opcode dispatch is so central to performance that it seems justified.

**Comment by fatcerberus** [comment]
I may have found a regression with this change, I have to do a bit more testing to see if I can reproduce it.

**Comment by svaarala** [comment]
What was the outward effect?

**Comment by fatcerberus** [comment]
The one thing I did notice was, for my pause-on-error change, when stopped at an error GetCallStack would sometimes report the wrong line number and PC (the beginning of a function instead of the correct place).

**Comment by svaarala** [comment]
Things like that will be incorrect when a "sync" is missing (copying the local curr_pc to `act->pc`). I think I'll find this one simply by code reading.

But was this the regression you were referring to or was it something else?

**Comment by svaarala** [comment]
I think there's at least a sync missing in the case where the bytecode executor triggers a side effect like a getter call. It won't cause any harm other than the `act->pc` being incorrect so tracebacks etc will look wrong.

I'll confirm and fix that shortly.

**Comment by fatcerberus** [comment]
The specific scenario where I noticed this was where a Duktape/C function throws an error and the debugger pauses before longjmp, the reported PC for the ecma function directly below is incorrect

**Comment by svaarala** [comment]
I can replicate this behavior by using a breakpoint in e.g. a getter function and resuming until the breakpoint is hit. It can't be reproduced by stepping because that will continuously sync the PC. That's why I missed this in my basic debugger test before merge.

Fix and bug testcase coming shortly. Thanks!

**Comment by fatcerberus** [comment]
What's weird is that the "enter debugger now" helper I added for #286 already includes a sync before the interrupt, so I'm not actually sure why this happens.

**Comment by svaarala** [comment]
The missing sync was for a non-ecma-to-ecma call, so that the PC of the calling activation would be outdated. The sync you added is for the *current* activation?

**Comment by fatcerberus** [comment]
Ah, yes, that makes sense. :)

**Comment by svaarala** [comment]
Ok, fix (and bug test) is in master, here's a snapshot: http://duktape.org/snapshots/duktape-1.2.99-20150826172312-v1.2.0-401-g8b5bcca.tar.xz

**Comment by fatcerberus** [comment]
I have to go to work right now, but I'll test out the snapshot tonight and let you know.

**Comment by svaarala** [comment]
Sure, no problem - I think the fix should work for at least that GetCallStack issue, but it's of course best if you can verify it.

**Comment by fatcerberus** [comment]
The snapshot fixed the issue.  If you're curious, the specific issue I discovered turned out to have nothing to do with ecmascript-to-native calls, the bug was in `new` handling:

```javascript
// line 1
(function() {
    print("The 8:12 pig was here!");
    function f() { print(Duktape.act(-3).lineNumber); }
    new f();  // prints 3, should print 5
})();
```

But whatever you did appears to have fixed this case too.

**Comment by svaarala** [comment]
It was actually related: the fix was in `handle_call()` and `new` uses `handle_call()` even when calling an Ecmascript constructor.

**Comment by fatcerberus** [comment]
Just triggered an assert that might be related to this (without assertions, this ends in a segfault):
```
PANIC 54: assertion failed: curr_pc < DUK_HCOMPILEDFUNCTION_GET_CODE_END(thr->heap, fun) (duk_js_executor.c:2266) (calling abort)
```

**Comment by fatcerberus** [comment]
The value of curr_pc at the time according to MSVC:
`curr_pc = 0x000000dfab127cb8 {4261281277}`

This value is consistent from run to run.

**Comment by svaarala** [comment]
Any idea what's going on when this happens?

**Comment by fatcerberus** [comment]
Well, here's my native callstack for starters:
```
 	engine64.exe!issue_debug_notification(const wchar_t * const message) Line 125	C++
 	engine64.exe!__acrt_report_runtime_error(const wchar_t * message) Line 142	C++
 	engine64.exe!abort() Line 61	C++
 	engine64.exe!duk_default_panic_handler(int code, const char * msg) Line 127	C
>	engine64.exe!duk_js_execute_bytecode(duk_hthread * exec_thr) Line 2266	C
 	engine64.exe!duk_handle_call(duk_hthread * thr, int num_stack_args, unsigned int call_flags) Line 1496	C
 	engine64.exe!duk_call(duk_hthread * ctx, int nargs) Line 61	C
 	engine64.exe!run_script(script * script, bool allow_reentry) Line 189	C
 	engine64.exe!render_map() Line 1284	C
 	engine64.exe!js_RenderMap(duk_hthread * ctx) Line 2856	C
 	engine64.exe!duk_handle_call(duk_hthread * thr, int num_stack_args, unsigned int call_flags) Line 1383	C
 	engine64.exe!duk_js_execute_bytecode(duk_hthread * exec_thr) Line 3408	C
 	engine64.exe!duk_handle_call(duk_hthread * thr, int num_stack_args, unsigned int call_flags) Line 1496	C
 	engine64.exe!duk_call_method(duk_hthread * ctx, int nargs) Line 84	C
 	engine64.exe!duk_bi_function_prototype_apply(duk_hthread * ctx) Line 206	C
 	engine64.exe!duk_handle_call(duk_hthread * thr, int num_stack_args, unsigned int call_flags) Line 1383	C
 	engine64.exe!duk_js_execute_bytecode(duk_hthread * exec_thr) Line 3408	C
 	engine64.exe!duk_handle_call(duk_hthread * thr, int num_stack_args, unsigned int call_flags) Line 1496	C
 	engine64.exe!duk_call(duk_hthread * ctx, int nargs) Line 61	C
 	engine64.exe!run_script(script * script, bool allow_reentry) Line 189	C
 	engine64.exe!update_map_engine(bool is_main_loop) Line 1397	C
 	engine64.exe!js_MapEngine(duk_hthread * ctx) Line 1546	C
 	engine64.exe!duk_handle_call(duk_hthread * thr, int num_stack_args, unsigned int call_flags) Line 1383	C
 	engine64.exe!duk_js_execute_bytecode(duk_hthread * exec_thr) Line 3408	C
 	engine64.exe!duk_handle_call(duk_hthread * thr, int num_stack_args, unsigned int call_flags) Line 1496	C
 	engine64.exe!duk_pcall(duk_hthread * ctx, int nargs) Line 138	C
 	engine64.exe!main(int argc, char * * argv) Line 337	C
 	[External Code]	
```

**Comment by fatcerberus** [comment]
Hm, I wonder if I could use Duktape's JS debugger to narrow things down?  There are a few levels of native calls in there though that might make it more difficult, though... and that's assuming single-stepping doesn't mask the bug.

**Comment by svaarala** [comment]
The most likely reason is that some PC sync operation uses an out-of-date thr->ptr_curr_pc somewhere. Just to narrow things down, do you use coroutines?

**Comment by svaarala** [comment]
I think I'll make a small change where thr->ptr_curr_pc is set to NULL whenever any activation states change: otherwise a PC sync operation would copy the previous curr_pc value into another activation.

**Comment by fatcerberus** [comment]
No coroutines, no.  But some weird corruption is definitely going on here.

Specifically, this comes up when starting the Spectacles battle engine (that being the game I wrote minisphere for).  The first battle I tested lead to the segfault and failed assert mentioned above.  I tried a different battle, and I'm SO glad I implemented pause-on-throw now... Duktape threw a nonsensical "invalid index" error (valid script filename but line number of 0), allowing me to view the Ecmascript callstack (which is quite deep!) and local variables.

With that information, I found the function responsible for the segfault:

```
	if (!this.isVisible && this.damages.length == 0 && this.healings.length == 0) {
		return;
	}
	this.sprite.blit(this.x, this.y, this.opacity * 255);
	for (var i = 0; i < this.damages.length; ++i) {
		var text = this.damages[i].text;
		var x = this.x + 16 - this.messageFont.getStringWidth(text) / 2;
		for (var i2 = 0; i2 < text.length; ++i2) {
			var yName = 'y' + i2.toString();
			var y = this.y + this.damages[i][yName];
			var color = this.damages[i].color !== null ? this.damages[i].color
				: CreateColor(255, 255, 255, 255);
			DrawTextEx(this.messageFont, x, y, text[i2], color, 1);
			x += this.messageFont.getStringWidth(text[i2]);
		}
	}
	for (var i = 0; i < this.healings.length; ++i) {
		var y = this.y + this.healings[i].y;
		var color = this.healings[i].color !== null ? this.healings[i].color : CreateColor(64, 255, 128, 255);
		var textColor = BlendColors(color, color);
		textColor.alpha *= this.healings[i].alpha / 255;
		DrawTextEx(this.messageFont, this.x + 16, y, this.healings[i].amount, textColor, 1, 'center');
	}
```

And the line which called it:
`this.actors[type][i].render();`

**Comment by fatcerberus** [comment]
Worth noting is that the local variables are all listed as being set to `undefined`.  Which means the `render()` function probably didn't even start executing yet.

**Comment by fatcerberus** [comment]
Forgive me if I seem a little overexcited here; I probably enjoy debugging more than would generally be considered healthy. :stuck_out_tongue_closed_eyes:

**Comment by svaarala** [comment]
:-)

**Comment by fatcerberus** [comment]
What I don't understand is why the value of the corrupt `curr_pc` isn't changing from run to run.  If it were using an outdated value (from a previous activation, say), the actual value should still change as memory allocation isn't exactly known for being deterministic, even with virtual addressing.  It looks more to me like `curr_pc` is being clobbered.

**Comment by svaarala** [comment]
This debug log just before the assert error might help shed some light:

```
DUK_D(DUK_DPRINT("curr_pc=%p base=%p end=%p &curr_pc=%p thr->ptr_curr_pc=%p",
    (void *) curr_pc,
    (void *) DUK_HCOMPILEDFUNCTION_GET_CODE_BASE(thr->heap, fun),
    (void *) DUK_HCOMPILEDFUNCTION_GET_CODE_END(thr->heap, fun),
    (void *) &curr_pc,
    (void *) thr->ptr_curr_pc));
```

**Comment by fatcerberus** [comment]
Okay, I'll add that debug print now.  By the way, the error is reproducible in a 32-bit x86 build as well, and will wonders never cease:
curr_pc = 0x0671d87c {4261281277}

Still the same!  Of course I realize now that the value in braces is the result of _dereferencing_ the pointer, but still intriguing that it's the same value.

**Comment by svaarala** [comment]
I'm trying to figure out what your game code does differently from the testcases: all testcases pass with assertions enabled. One difference is that your calling C side is obviously more complex than the "duk" command line tool, and that there are more Duktape/C functions involved.

I've added setting ptr_curr_pc to NULL whenever activations are changed (when doing call setup and when throwing errors). This is so that while the pointer itself may be valid, it's no longer related to the topmost activation's curr_pc pointer so syncing would be incorrect. I'm not sure if this is what's happening in your case; but here's a snapshot with the change: http://duktape.org/snapshots/duktape-1.2.99-20150827100951-v1.2.0-403-gf0535a3.tar.xz

**Comment by svaarala** [comment]
That decimal value in hex is `0xfdfdfdfd` by the way.

**Comment by fatcerberus** [comment]
Lovely, that debug print generates so much output that the game never gets started because the console is bogged down printing stuff.

**Comment by fatcerberus** [comment]
Got it, had to be clever with a conditional:
[D] duk_js_executor.c:2271 (unknown): curr_pc=00000094B45DE5B8 base=00000094B500
3298 end=00000094B50035BC &curr_pc=00000094B13E9F58 thr->ptr_curr_pc=00000094B13
E9F58

**Comment by fatcerberus** [comment]
As for that 0xfdfdfdfd value, I found this bit of info:
"0xFD 	Fence Memory 	Also known as "no mans land." This is used to wrap the allocated memory (like surrounding it with fences) and is used to detect indexing arrays out of bounds."

Interesting.

**Comment by fatcerberus** [comment]
Unfortunately it's 3:30am and I need some sleep, so further bug-hunting will have to wait. :)

Looking at that debug print though, it looks like the PC someone ended up just outside the bytecode.  Which is consistent with the "fence memory" thing.

**Comment by svaarala** [comment]
No worries, I'll try to figure this out. Thanks for the help!

**Comment by svaarala** [comment]
I think I'll add a `thr->idx_curr_act` to accompany `thr->ptr_curr_pc` which identifies the activation the ptr_curr_pc is related to. This makes syncing always correct even when there are activation changes, so it'll be much easier to get the syncing correct.

**Comment by svaarala** [comment]
With the hint from `apply()` I found a code snippet which causes a valgrind detected error:

```js
function f() {
    print('f');
    function g() {
        print('g');
        throw new Error('myerror');
    }
    g();
}

try {
    // Calling as f() does not trigger the error, but calling through
    // apply() does.  The internal difference is that apply() goes through
    // duk_handle_call().
    f.apply();
} catch (e) {
    print(e);
}
```

**Comment by svaarala** [comment]
Ok, I think I found one error at least: the bytecode executor setjmp catchpoint would do a curr_pc sync which doesn't work if (1) there are multiple executor instances in the C call stack (apply does that above), (2) the error propagates from one executor setjmp catchpoint to the next without being caught in the middle. The outer setjmp catchpoint would then do a curr_pc sync which would reference the inner executor instance whose C stack frame is now invalid.

The cleanest approach is to sync curr_pc always before a throw (longjmp) and never in a setjmp catcher. I'll work up a fix to this effect and post a snapshot shortly.

**Comment by svaarala** [comment]
Removed the curr_pc sync from the longjmp handler which fixes the bug testcase. Merged to master so should appear in the latest snapshots in about an hour.

**Comment by svaarala** [comment]
Here's a snapshot from master including the (potential) fix: http://duktape.org//snapshots/duktape-1.2.99-20150827130003-v1.2.0-408-gd3abfae.tar.xz

**Comment by fatcerberus** [comment]
Oh good, I'll test it out a bit later.

Funny you mention apply() though.... I believe there was an apply somewhere in the JS callstack that one time I got it to throw instead of segfaulting.

**Comment by fatcerberus** [comment]
Nope, I still get the segfault in the exact same spot even with the snapshot.

**Comment by fatcerberus** [comment]
Found the exact statement where the segfault happens:
```javascript
this.sprite.blit(this.x, this.y, this.opacity * 255);
```

`blit()` is implemented on the native side.  For some reason execution of this statement sends `curr_pc` out of bounds...

**Comment by fatcerberus** [comment]
I'm about to try something very unorthodox, stand by. :)

**Comment by fatcerberus** [comment]
Okay, the bug appears to be in call return handling.  The engine segfaults when running this function:

```javascript
	this.blit = function(x, y, alpha)
	{
		alpha = alpha !== void null ? alpha : 255;
		
		this.spriteset.images[this.spriteset.directions[this.directionID].frames[this.frameID].index]
			.blitMask(x + this.xOff, y + this.yOff, CreateColor(255, 255, 255, alpha));
		//print("*munch*");
	};
```

However, the function itself completes--the crash happens after the function returns but before the calling activation resumes.

Notice that commented-out `print` call I added?  Here's where things get weird: If I uncomment that, everything runs normally.

**Comment by svaarala** [comment]
So is this blit a replacement for the native call (you said blit was implemented natively)?

**Comment by fatcerberus** [comment]
No, it turns out the blit being called actually was this one, I had forgotten about my custom SpriteImage class.

**Comment by svaarala** [comment]
Do you get the same error with this snapshot: http://duktape.org/snapshots/duktape-1.2.99-20150825070003-v1.2.0-382-g64560f8.tar.xz

(It has the thr->curr_pc dispatch which is similar but a bit slower.)

**Comment by svaarala** [comment]
What about `.blitMask()`, is that a native call?

**Comment by fatcerberus** [comment]
No crash with the snapshot.

`.blitMask` is native, yes.  And `.blit()` isn't on the callstack at the time of the crash.

**Comment by svaarala** [comment]
Kind of hard to pinpoint the problem without a repro case, but what about this: http://duktape.org/snapshots/duktape-1.2.99-20150827185416-v1.2.0-408-gd3abfae-dirty.tar.xz

**Comment by fatcerberus** [comment]
Best I can tell, curr_pc is being restored incorrectly during call return and being placed in no-man's-land.  I'm trying as hard as I can to reproduce it without minisphere-specifics, but not having much luck so far. :(

**Comment by svaarala** [comment]
I think the difference between our environments is that my setup doesn't have "stack wiping" so that dangling references to unwound stack frames will luck out and work (really well in fact because all tests pass). I'll look up some stack debugging options to see if I can get similar behavior.

**Comment by svaarala** [comment]
Anyway, that last snapshot is worth a try :)

**Comment by fatcerberus** [comment]
It segfaults in a release build for me as well, so I think it's something a bit more insidious than a simple dangling pointer.  Anyway, I'll try the snapshot in a bit.

**Comment by svaarala** [comment]
What compiler and options are you using BTW? I can try to reproduce this by running the tests on Windows instead.

**Comment by fatcerberus** [comment]
Compiler is MSVC 2015.

Options are:
`
#define DUK_OPT_TARGET_INFO  ENGINE_NAME
#define DUK_OPT_FASTINT
#define DUK_OPT_DEBUGGER_SUPPORT
#define DUK_OPT_INTERRUPT_COUNTER
#define DUK_OPT_DEBUGGER_FWD_PRINTALERT
`

I can reproduce the crash without the debugger options too, though.

**Comment by svaarala** [comment]
Ah, I meant the compiler options - optimization level, etc.

**Comment by fatcerberus** [comment]
Oh.  I'm essentially using stock settings, the relevant customizations are:
Optimize for speed (/O2)
Static multithread CRT (/MT)

**Comment by fatcerberus** [comment]
Still crashes with that last snapshot.

**Comment by svaarala** [comment]
VS2013 x64 with /O2 and /MT seems to work fine, so I haven't been able to reproduce this so far.

I'll give this a little more time, but if there's no way to pin this down, I'll revert this and perhaps postpone it to 1.4.

**Comment by fatcerberus** [comment]
Since I can reproduce this on command in the minisphere environment, I'll do a bit of stepping through Duktape and see if I can turn up anything.

**Comment by fatcerberus** [comment]
Something interesting though: print() calls inside of .blit() don't get redirected to the debugger, but so show up in the console output.

**Comment by svaarala** [comment]
Do you only get this error from .blit() by the way? What happens if you make .blitMask() (the C function) just return and do nothing? (It's unlikely this would be an application problem but, just to rule things out.)

**Comment by fatcerberus** [comment]
Empty .blitMask() doesn't make a difference.  So something about the construction of the .blit() function is triggering a bug during call return (or even during the call, given the print() oddity), I just don't know what.

But yeah, calling .blit() is the only way I can reproduce it so far.  I've been trying all sorts of odd combinations of Ecmascript and C calls ever since to try to reproduce it in isolation, but nothing.  Next I will run the whole thing through Application Verifier (like valgrind but not as good) to see if I can pick up any memory-unsafe shenanigans.

**Comment by svaarala** [comment]
This is the bytecode for .blit() BTW :-)

```
; c0 "spriteset"
; c1 "images"
; c2 "directions"
; c3 "directionID"
; c4 "frames"
; c5 "frameID"
; c6 "index"
; c7 "blitMask"
; c8 "xOff"
; c9 "yOff"
; c10 "CreateColor"

00000 000100f3   LDUNDEF    r4
00001 020080e8   SNEQ       r3, r2, r4
00002 0000c06d   IF         true, r3
00003 800000ae   JUMP       6 (+3)
00004 000080c0   LDREG      r3, r2
00005 8000006e   JUMP       7 (+2)
00006 803fc0c3   LDINT      r3, 255
00007 0000c080   LDREG      r2, r3
00008 0000c0b3   LDTHIS     r3
00009 8000c0d5   GETPROP    r3, r3, c0
00010 8080c0d5   GETPROP    r3, r3, c1
00011 000100b3   LDTHIS     r4
00012 80010115   GETPROP    r4, r4, c0
00013 81010115   GETPROP    r4, r4, c2
00014 000140b3   LDTHIS     r5
00015 81814155   GETPROP    r5, r5, c3
00016 02810115   GETPROP    r4, r4, r5
00017 82010115   GETPROP    r4, r4, c4
00018 000180b3   LDTHIS     r6
00019 82818195   GETPROP    r6, r6, c5
00020 03010115   GETPROP    r4, r4, r6
00021 83010115   GETPROP    r4, r4, c6
00022 0200c0d5   GETPROP    r3, r3, r4
00023 0000c1c0   LDREG      r7, r3
00024 0001c202   LDCONST    r8, c7
00025 0401c1d8   CSPROP     r7, r7, r8
00026 000240b3   LDTHIS     r9
00027 84024255   GETPROP    r9, r9, c8
00028 0480025a   ADD        r9, r0, r9
00029 000280b3   LDTHIS     r10
00030 84828295   GETPROP    r10, r10, c9
00031 0500429a   ADD        r10, r1, r10
00032 004282d2   CSVAR      r11, c10
00033 803fc343   LDINT      r13, 255
00034 803fc383   LDINT      r14, 255
00035 803fc3c3   LDINT      r15, 255
00036 00008400   LDREG      r16, r2
00037 0202c030   CALL       0x0, r11, 4
00038 0181c030   CALL       0x0, r7, 3
00039 0000006f   RETURN     0x1, r0          ; fast_return
```

**Comment by fatcerberus** [comment]
Okay, I may have a lead: I attached the MSVC debugger to engine64.exe, performed a StepOver in Duktape on the .blit() call, and it generated a debug interrupt BEFORE .blit() returned.  Guess what?  curr_pc is already in no-man's-land at that point, which means issuing a Resume is going to segfault.

**Comment by fatcerberus** [comment]
The segfault is indeed in return handling.  Unfortunately MSVC won't follow the longjmp for me to see exactly where it all goes wrong.

**Comment by svaarala** [comment]
If blit() is called from Ecmascript code, what should happen is:

- The RETURN opcode sets up the longjmp type as DUK_LJ_TYPE_RETURN, syncs curr_pc, and then longjmps.
- The longjmp handler switches based on the longjmp type, detects that there's an Ecmascript caller and that we can reuse the executor C call. It then unwinds the stacks, reconfigures the value stack, and returns DUK__LONGJMP_RESTART which causes a goto reset_setjmp_catchpoint.
- The reset_setjmp_catchpoint labels creates a new setjmp catchpoint, rechecks debugger state, and resets curr_pc before restarting execution.

**Comment by svaarala** [comment]
What happens if you StepInto the .blit() function and then StepOver() there? Does the problem occur when you've executed al statements and are about to return?

**Comment by svaarala** [comment]
The longjmp() you're trying to follow is most likely landing in the executor setjmp catchpoint, so maybe you could add a separate breakpoint there?

- https://github.com/svaarala/duktape/blob/master/src/duk_js_executor.c#L2049

**Comment by fatcerberus** [comment]
Yes, performing a StepOver at the end of the .blit() function causes the debugger to pause just before returning to the caller.  Performing another step (or resuming) asserts.

And yes, I did exactly that (the breakpoint on line 2049) . What I ascertained is that the previous activation (the one that called .blit()) is for some reason not being restored.  Duktape still believes it's executing .blit() (`*(act->curr_pc) == 0x6f`), and attempts to continue past the return, leading to a crash.

**Comment by svaarala** [comment]
One thing to rule out might be ensuring this is not a setjmp related issue. That'd be ruled out if you changed:

- https://github.com/svaarala/duktape/blob/master/src/duk_js_executor.c#L1978

to:

```
duk_instr_t * volatile curr_pc;
```

(Ignore the warnings related to casts.)

This shouldn't be an issue though: while the value of curr_pc is not guaranteed for the setjmp() catch path, we've synced the curr_pc before that and shouldn't actually be reading it in the catch path.

**Comment by fatcerberus** [comment]
Yeah, making curr_pc volatile didn't help any.

**Comment by svaarala** [comment]
Did you step through the longjmp handling? If the caller to .blit() is an Ecmascript call (and is handled normally using an Ecmascript-to-Ecmascript call), the expected path is roughly:

- Detect longjmp type: https://github.com/svaarala/duktape/blob/master/src/duk_js_executor.c#L1109
- Notice that we're returning to bytecode executor which reused its C stack frame: https://github.com/svaarala/duktape/blob/master/src/duk_js_executor.c#L1154
- Unwinding is done, returning and restarting: https://github.com/svaarala/duktape/blob/master/src/duk_js_executor.c#L1176

At that last line `thr->callstack_top` should have been decremented by one compared to the second step (line 1154). The executor restart should then pick up executing the caller.

**Comment by fatcerberus** [comment]
Okay, I found something suspicious:
```c
thr->heap->call_recursion_depth = entry_call_recursion_depth;
```

This is the code path leading to the assert, and before the above line is executed:
```
thr->heap->call_recursion_depth = 1
entry_call_recursion_depth = 1
```

That doesn't seem right.

**Comment by fatcerberus** [comment]
`callstack_top` indeed goes from 2 to 1 between the second and third steps.

So the return itself is fine, but the callstack is corrupt, because:
```
*(act->curr_pc - 2) = 181C030
```

Which is the opcode right before the return from .blit().

**Comment by svaarala** [comment]
But has `act` been looked up again after the unwind? The executor should re-lookup `act = thr->callstack + thr->callstack_top - 1` -- is this the `act` you refer to above?

I stubbed out stuff from your example: does this work for you as a standalone script?

```
function Image() {
}
Image.prototype.blitMask = function (x, y, color) {
    print('blitMask', x, y, color);
};
var myImage = new Image();

function CreateColor() { return 123; }

this.spriteset = {
    directions: [ { frames: [ { index: 0 } ] } ],
    images: [ myImage ],
};
this.directionID = 0;
this.frameID = 0;
this.xOff = 10;
this.yOff = 50;

this.blit = function(x, y, alpha)
    {
        alpha = alpha !== void null ? alpha : 255;

        this.spriteset.images[this.spriteset.directions[this.directionID].frames[this.frameID].index]
            .blitMask(x + this.xOff, y + this.yOff, CreateColor(255, 255, 255, alpha));
        //print("*munch*");
    };

this.blit(1,2,3);
print('done');
```

**Comment by svaarala** [comment]
Oh btw, does any value in .blit() have a finalizer? :-)

**Comment by svaarala** [comment]
The reason I ask is that the unwind handling may have side effects which causes side effects, including function calls, and the state may not be correct to do that because `thr->ptr_curr_pc` has not been NULLed. A function call will then sync the PC to the incorrect activation.

If this is what's happening, changing the end of the DUK_OP_RETURN handling to this should help:

```
duk_hthread_sync_currpc(thr);
thr->ptr_curr_pc = NULL;
duk_err_longjmp(thr);
DUK_UNREACHABLE();
```

A similar change will be needed in other code paths, but maybe if you try this we'd see if side effects during unwind are indeed the issue.

**Comment by fatcerberus** [comment]
The Image objects which `.blitMask()` is called on have finalizers, but I don't see why they would be called, they are still reachable through the SpriteImage object.

Your stub code causes a DIFFERENT assert, and only with the Duktape debugger attached:
```
blitMask 11 52 123
done
PANIC 54: assertion failed: DUK_HOBJECT_HAS_COMPILEDFUNCTION((duk_hobject *) fun) (duk_js_executor.c:1729) (calling abort)
```

**Comment by fatcerberus** [comment]
Nulling PC on return fixes real `.blit()`, your testcase still asserts as above though.

**Comment by svaarala** [comment]
Ok, I think we're finally getting to the real cause here: side effects during unwinding can cause a function call to happen and if `thr->ptr_curr_pc` is set at that point, call handling will "sync pc" using that pointer and the topmost activation -- which may be incorrect. That's why the PC in the outer function was that of the inner function.

I had a temporary branch where I added an explicit callstack index to accompany the `ptr_curr_pc` so that the syncing would always go to the right activation; this would then work even with side effects. The other alternative is just to work through all the call paths and ensure every path uses a "sync-and-null".

**Comment by fatcerberus** [comment]
So what's with that `DUK_HOBJECT_HAS_COMPILEDFUNCTION()` assert?  Same root cause?  The code you posted works fine standalone, but put a Duktape/C call after it and it asserts (but again, only with attached debugger).

**Comment by svaarala** [comment]
I think I managed to create a test case to exercise the current problem:

```js
function finalizer() {
    print('obj finalized');
}

function func() {
    var obj = {};
    Duktape.fin(obj, finalizer);

    print('returning');
}

function test() {
    func();
    func();
}

try {
    test();
} catch (e) {
    print(e.stack || e);
}
```

The unwind of func() causes the finalizer call, which then causes a break on the second func() call. This breaks for me too, and is fixed by the change to NULL the ptr_curr_pc.

**Comment by svaarala** [comment]
Here's a snapshot with the ptr_curr_pc NULLing: http://duktape.org/snapshots/duktape-1.2.99-20150827221115-v1.2.0-408-gd3abfae-dirty.tar.xz

**Comment by fatcerberus** [comment]
I'll try the snapshot tonight.  I'm glad the cause was found though, this was quite a puzzler and I have a splitting headache now. :)

**Comment by svaarala** [comment]
Me too, thanks for being awesomely patient :-)

**Comment by svaarala** [comment]
Here's the fix branch: https://github.com/svaarala/duktape/compare/fix-currpc-longjmp-sync

And here's a snapshot built from that: http://duktape.org/snapshots/duktape-1.2.99-20150828014009-v1.2.0-412-ga09c0da.tar.xz

If you could retest the debugger issue with this snapshot that'd be useful for figuring out the cause.

(I'm off to sleep soonish, it's been an interesting but exhausting bug huntery :-)

**Comment by fatcerberus** [comment]
Tried both snapshots, both fix the original bug as well as the HAS_COMPILEDFUNCTION assertion, so apparently that was just a side-effect of the same bug.  So just to clarify, was it indeed the case that curr_pc was being synced to the wrong activation?

**Comment by fatcerberus** [comment]
Does the fix affect performance at all, out of curiosity?  I'd hate for you to have gone to all the trouble to implement this just to end up nullifying all the performance benefits with the fix...

**Comment by svaarala** [comment]
The changes are all related to figuring out where to (a) sync curr_pc back to act->curr_pc and (b) when to set/reset thr->ptr_curr_pc so everything works out correctly. None of these are on the opcode dispatch hot path so there should be no direct effect on performance of the hot path.

**Comment by svaarala** [comment]
It's actually quite hard to get exact, useful performance results for opcode dispatch performance because it depends on such a lot of things. For example, an innocent change may have a +/- 5% impact with -O2, probably caused by things like register allocation / spilling decisions, code alignment, etc.

Anyway, the overall effect of bytecode executor changes between v1.2.3 and this fix branch is quite positive. For example, testing `tests/perf/test-reg-load.js` which is very dispatch heavy using:

- clang -O3
- DUK_OPT_FASTINT
- no debugger/interrupt support

Has the results:

```
# Duktape v1.2.3
$ sudo nice -20 python util/time_multi.py --count 10 --mode all --sleep 5 --verbose ./duk.O3clang.123 tests/perf/test-reg-load.js 
Running: 4.070000 4.060000 4.060000 4.060000 4.050000 4.050000 4.060000 4.070000 4.060000 4.060000
min=4.05, max=4.07, avg=4.06, count=10: [4.07, 4.06, 4.06, 4.06, 4.05, 4.05, 4.06, 4.07, 4.06, 4.06]

# This fix branch
$ sudo nice -20 python util/time_multi.py --count 10 --mode all --sleep 5 --verbose ./duk.O3clang.currpc tests/perf/test-reg-load.js 
Running: 2.760000 3.020000 2.820000 2.970000 2.740000 2.760000 2.750000 2.820000 2.750000 2.750000
min=2.74, max=3.02, avg=2.81, count=10: [2.76, 3.02, 2.82, 2.97, 2.74, 2.76, 2.75, 2.82, 2.75, 2.75]
```

The minimum time is 4.05 seconds for Duktape v1.2.3 and 2.74 seconds for this branch, which is well in line for a ~25% dispatch improvement.

**Comment by svaarala** [comment]
By the way, your code uses `void null` for `undefined` which is something I'll add to performance notes. Because `undefined` is not a literal it causes a `GETVAR` to be emitted, while `void null` (or `void 1` etc) cause a `LDUNDEF`. You can see from the bytecode that nothing extra gets emitted which is nice :)

**Comment by fatcerberus** [comment]
So long story short, Duktape 1.3.0 will be a very noticeable performance improvement over 1.2.x, very nice.

As for `void null`, that was actually code I wrote back when I was using the original Sphere, which used an ES3 engine (SpiderMonkey 1.5).  Since the value of `undefined` can be reassigned in ES3, `void null` was a way to _ensure_ I got a real undefined back.  The fact that it represents a performance improvement is a nice bonus. :)

**Comment by svaarala** [comment]
With the usual caveat that micro benchmarks don't represent application performance :)

**Comment by fatcerberus** [comment]
Perhaps not, but I can say that I've seen a noticeable performance improvement in minisphere since I started using the latest masters.  My game starts up much faster than it used to of late. ;-)

**Comment by fatcerberus** [comment]
So, how many bug testcases have been written as a result of this issue now?  I'm legitimately curious.

**Comment by svaarala** [comment]
There were three actual bug testcases reproducing some aspects of the root cause, and one testcase for the stubbed out blit().

**Comment by svaarala** [comment]
I think I figured out a torture test to shake out these side effect related bugs: do a dummy function call for every refzero queue free operation (simulating the case that every object has a finalizer). This should've exposed the finally-related finalizer issues too. I'll see about that in a few days when I've got some other things out of the way.

**Comment by svaarala** [comment]
I'll close this issue now unless there's some fallout to work on. For new fallout it's probably best to open a separate issue as this is already quite long :)

**Comment by svaarala** [comment]
Oh and that refzero torture issue is #304.
